awx_collection/plugins/modules/tower_job_template.py

```diff
@@ -12,409 +13,562 @@
-except ImportError:
-    pass
-
-
-def update_fields(module, p):
-    '''This updates the module field names
-    to match the field names tower-cli expects to make
-    calling of the modify/delete methods easier.
-    '''
-    params = p.copy()
-    field_map = {
-        'fact_caching_enabled': 'use_fact_cache',
-        'ask_diff_mode': 'ask_diff_mode_on_launch',
-        'ask_extra_vars': 'ask_variables_on_launch',
-        'ask_limit': 'ask_limit_on_launch',
-        'ask_tags': 'ask_tags_on_launch',
-        'ask_skip_tags': 'ask_skip_tags_on_launch',
-        'ask_verbosity': 'ask_verbosity_on_launch',
-        'ask_inventory': 'ask_inventory_on_launch',
-        'ask_credential': 'ask_credential_on_launch',
-        'ask_job_type': 'ask_job_type_on_launch',
-        'diff_mode_enabled': 'diff_mode',
-        'concurrent_jobs_enabled': 'allow_simultaneous',
-        'force_handlers_enabled': 'force_handlers',
-    }
-
-    params_update = {}
-    for old_k, new_k in field_map.items():
-        v = params.pop(old_k)
-        params_update[new_k] = v
-
-    extra_vars = params.get('extra_vars')
-    extra_vars_path = params.get('extra_vars_path')
-
-    if extra_vars:
-        params_update['extra_vars'] = [json.dumps(extra_vars)]
-
-    elif extra_vars_path is not None:
-        params_update['extra_vars'] = ['@' + extra_vars_path]
-        module.deprecate(
-            msg='extra_vars_path should not be used anymore. Use \'extra_vars: "{{ lookup(\'file\', \'/path/to/file\') | from_yaml }}"\' instead',
-            version="3.8"
-        )
-
-    params.update(params_update)
-    return params
-
-
-def update_resources(module, p):
-    params = p.copy()
-    identity_map = {
-        'project': 'name',
-        'inventory': 'name',
-        'credential': 'name',
-        'vault_credential': 'name',
-    }
-    for k, v in identity_map.items():
-        try:
-            if params[k]:
-                key = 'credential' if '_credential' in k else k
-                result = tower_cli.get_resource(key).get(**{v: params[k]})
-                params[k] = result['id']
-            elif k in params:
-                # unset empty parameters to avoid ValueError: invalid literal for int() with base 10: ''
-                del(params[k])
-        except (exc.NotFound) as excinfo:
-            module.fail_json(msg='Failed to update job template: {0}'.format(excinfo), changed=False)
-    return params
-
-
 def main():
+    # Any additional arguments that are not fields of the item can be added here
     argument_spec = dict(
-        name=dict(required=True),
-        description=dict(default=''),
-        job_type=dict(choices=['run', 'check']),
-        inventory=dict(default=''),
-        project=dict(required=True),
-        playbook=dict(required=True),
+        name=dict(required=True, type='str'),
+        new_name=dict(required=False, type='str'),
+        description=dict(required=False, type='str', default=''),
```

Comment:
I'm moderately concerned about all the defaults here, mostly in the case of modifying a job_template, if we have a default and the user doesn't specify a value for that parameter we'll overwrite that parameter with the default, we do need defaults for a lot of these things, but only for new objects. I think we ran into this issue with a different module, i forget which one, maybe user

Reply:
we've discussed this in some other places too. I'm in agreement, my stance is that `state` should have a default, and the rest can go.